### PR TITLE
OAK-6911 - Provide a way to tune inline size while storing binaries

### PR DESCRIPTION
**Pull request by dulceanu**

### CODE REVIEW

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/FileStoreBuilder.java [hunk 1]
```diff
@@ -396,6 +401,17 @@ public FileStoreBuilder withEagerSegmentCaching(boolean eagerSegmentCaching) {
...
         return this;
     }
+    
+    /**
+     * Sets the threshold under which binaries are inlined in data segments.
+     * @param binariesInlineThreshold the threshold
+     * @return this instance
+     */
+    public FileStoreBuilder withBinariesInlineThreshold(int binariesInlineThreshold) {
+        checkArgument(binariesInlineThreshold <= Segment.MEDIUM_LIMIT);
```
Comment (thomasmueller) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/FileStoreBuilder.java]: I would add:
checkArgument(binariesInlineThreshold >= 0);

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java [hunk 2]
```diff
@@ -132,14 +134,16 @@ public DefaultSegmentWriter(
             @NotNull SegmentIdProvider idProvider,
             @Nullable BlobStore blobStore,
             @NotNull WriterCacheManager cacheManager,
-            @NotNull WriteOperationHandler writeOperationHandler
+            @NotNull WriteOperationHandler writeOperationHandler,
+            int binariesInlineThreshold
```
Comment (mreutegg) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:140]: JavaDoc comment is missing for this parameter.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java [hunk 3]
```diff
@@ -132,14 +134,16 @@ public DefaultSegmentWriter(
...
     ) {
         this.store = checkNotNull(store);
         this.reader = checkNotNull(reader);
         this.idProvider = checkNotNull(idProvider);
         this.blobStore = blobStore;
         this.cacheManager = checkNotNull(cacheManager);
         this.writeOperationHandler = checkNotNull(writeOperationHandler);
+        this.binariesInlineThreshold = binariesInlineThreshold;
```
Comment (mreutegg) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:150]: The FileStoreBuilder checks this argument, but not this public constructor. Would it be better to move the check here?

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java [hunk 4]
```diff
@@ -638,11 +642,13 @@ private RecordId writeStream(@NotNull InputStream stream) throws IOException {
         }
 
         private RecordId internalWriteStream(@NotNull InputStream stream) throws IOException {
-            // Special case for short binaries (up to about 16kB):
+            // Special case for short binaries (up to about binariesInlineThreshold - 16kB default):
```
Comment (mreutegg) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java]: This comment confused me. I interpreted the dash as an arithmetic operation, but it's not, right? Maybe remove '- 16kB default' or separate with a semicolon.

File: oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	public TemporaryFolder folder = new TemporaryFolder(new File("target"));
+
+//	private TemporaryBlobStore blobStore = new TemporaryBlobStore(folder);
+//
+//	private TemporaryFileStore fileStore = new TemporaryFileStore(folder, blobStore, false);
+	
+//	@Rule
+//	public RuleChain rules = RuleChain.outerRule(folder).around(blobStore).around(fileStore);
+
+	private BlobStore createBlobStore(File dir) throws Throwable {
```
Comment (mreutegg) [inline oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java]: Throws clause is unnecessary.

File: oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+	@Rule
+	public TemporaryFolder folder = new TemporaryFolder(new File("target"));
+
+//	private TemporaryBlobStore blobStore = new TemporaryBlobStore(folder);
+//
+//	private TemporaryFileStore fileStore = new TemporaryFileStore(folder, blobStore, false);
+	
+//	@Rule
+//	public RuleChain rules = RuleChain.outerRule(folder).around(blobStore).around(fileStore);
```
Comment (mreutegg) [inline oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java]: Remove commented lines.

File: oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
+/*
```
Comment (mreutegg) [inline oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/BinariesInlineThresholdIT.java:1]: General comment about this file. Can you please convert tabs into spaces?

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java [hunk 8]
```diff
@@ -651,6 +657,17 @@ private RecordId internalWriteStream(@NotNull InputStream stream) throws IOExcep
...
+            
+            // handle case in which blob store is not configured and
+            // binariesInlineThreshold < Segment.MEDIUM_LIMIT
+            // store the binaries as small or medium sized value records 
+            
+            data = Arrays.copyOf(data, Segment.MEDIUM_LIMIT);
+            n += read(stream, data, n, Segment.MEDIUM_LIMIT - n);
+            
+            if (n < Segment.MEDIUM_LIMIT) {
+                return writeValueRecord(n, data);
```
Comment (mreutegg) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:673]: AFAICS, this line is not covered by a unit test.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java (hunk 2 above)
Comment (dulceanu) [reply oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:140]: Added missing JavaDoc comment.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java (hunk 3 above)
Comment (dulceanu) [reply oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:150]: Yes, I think you're right. I moved the checks here.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/FileStoreBuilder.java (hunk 1 above)
Comment (dulceanu) [reply oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/FileStoreBuilder.java]: I added that check as well, thanks for chiming in!

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java (hunk 4 above)
Comment (dulceanu) [reply oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java]: I agree, it was confusing. I clarified it in the latest commit.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java (hunk 8 above)
Comment (dulceanu) [reply oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/DefaultSegmentWriter.java:673]: I added `DefaultSegmentWriterTest#testValueRecord` to specifically test this part.

**Review by mreutegg — APPROVED** [review]
Looks good to me now.

**Comment by mreutegg** [comment]
Changes have been committed to SVN.

**Pull request closed without merging**